lib/auth/grpcserver.go

```diff
@@ -2944,18 +2961,42 @@ func (g *GRPCServer) GetTokens(ctx context.Context, _ *empty.Empty) (*types.Prov
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	provisionTokensV2 := make([]*types.ProvisionTokenV2, len(ts))
-	for i, t := range ts {
-		var ok bool
-		if provisionTokensV2[i], ok = t.(*types.ProvisionTokenV2); !ok {
-			return nil, trace.Errorf("encountered unexpected token type: %T", t)
+	provisionTokensV2 := make([]*types.ProvisionTokenV2, 0, len(ts))
+	for _, t := range ts {
+		v2Token, ok := t.V2()
+		if ok {
+			provisionTokensV2 = append(provisionTokensV2, v2Token)
+		} else {
+			g.Logger.
+				WithField("token_id", t.GetResourceID()).
+				Debug("GetTokens (V2) encountered ProvisionTokenV3 that could not convert to V2. Omitting.")
```

Comment:
Are you sure that omitting is a good call here? I understand that we want to be backward compatible, but if we find a token that cannot be converted, I'd argue that we should report an error to the client rather than silently ignore it. Otherwise, depending on the client version, you may get a different list of tokens which can be very confusing.